### PR TITLE
Plugins: Improve frontend loader cache

### PR DESCRIPTION
**Pull request by jackw**
**What is this feature?**

This PR changes the frontend plugin loader cache to apply caching query params to any js or css file that is loaded via SystemJS from the file system. It also makes changes so that nested plugins are loaded via their plugin id rather than their path (e.g. load via `plugins/k6-datasource/module.js` rather than `plugins/k6-app/datasource/module.js`) and if they lack a version in their plugin.json they will inherit the parent plugins version to be used as a cache buster during loading.

**Why do we need this feature?**

To improve the frontend loading and caching of plugin assets.

**Who is this feature for?**

Everyone

**Which issue(s) does this PR fix?**:

Fixes #

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

File: public/app/features/plugins/loader/cache.test.ts [hunk 1]
```diff
@@ -1,54 +1,64 @@
...
+}));
 
-describe('Plugin Cache', () => {
-  const now = 12345;
+describe('Cache Functions', () => {
+  describe('registerPluginInCache', () => {
+    it('should register a plugin in the cache', () => {
+      const plugin = { pluginId: 'plugin1', version: '1.0.0', isAngular: false };
+      registerPluginInCache(plugin);
+      expect(getPluginFromCache('plugin1')).toEqual(plugin);
```
Comment (leventebalogh) [inline public/app/features/plugins/loader/cache.test.ts:12]: **Nit:** although this test is fine, I think `getPluginFromCache()` will not always give back the same result as what you registered (it's gonna be a different object, in certain cases with different fields due to `encodeURI(version)`).

File: public/app/features/plugins/loader/cache.test.ts [hunk 2]
```diff
@@ -1,54 +1,64 @@
...
 
-  it('should append Date.now as cache flag if plugin is invalidated in buster', () => {
-    const slug = 'bubble-chart-3';
-    const version = 'v1.0.0';
-    const path = `/public/plugins/${slug}/module.js`;
-    const address = `http://localhost:3000/public/${path}.js`;
-
-    registerPluginInCache({ path, version });
-    invalidatePluginInCache(slug);
+  describe('resolveWithCache', () => {
```
Comment (leventebalogh) [inline public/app/features/plugins/loader/cache.test.ts:39]: **Nit:** maybe a name like `enrichURLWithCache` would be more descriptive.

File: public/app/features/plugins/loader/cache.test.ts [hunk 3]
```diff
@@ -1,54 +1,64 @@
...
+    it('should resolve URL with timestamp cache bust parameter', () => {
```
Comment (leventebalogh) [inline public/app/features/plugins/loader/cache.test.ts:40]:
```suggestion
    it('should resolve URL with timestamp cache bust parameter if plugin is not available in the cache', () => {
```

File: public/app/features/plugins/loader/cache.test.ts [hunk 4]
```diff
@@ -1,54 +1,64 @@
...
-  it('should also clear plugin settings cache', () => {
-    const slug = 'bubble-chart-3';
-    const version = 'v1.0.0';
-    const path = `/public/plugins/${slug}/module.js`;
-
-    const clearPluginSettingsCacheSpy = jest.spyOn(pluginSettings, 'clearPluginSettingsCache');
-
-    registerPluginInCache({ path, version });
-    invalidatePluginInCache(slug);
+  describe('getPluginFromCache', () => {
```
Comment (leventebalogh) [inline public/app/features/plugins/loader/cache.test.ts:53]: I would probably also add a test-case for `encodeURI(version)` part as well (to highlight in which scenarios it is necessary).

**Review by leventebalogh — APPROVED** [review]
LGTM 👍

**Review by xnyo — APPROVED** [review]
LGTM! Awesome work. Also tested locally with and without CDN and it works as expected. :rocket:

**Pull request merged**